File: code/TSP/GeneticAlgorithm.java
import java.util.Arrays;

public class GeneticAlgorithm {

    private int populationSize;
    private double mutationRate;
    private double crossoverRate;
    private int elitismCount;
    protected int tournamentSize;

    public GeneticAlgorithm(int populationSize, double mutationRate, double crossoverRate, int elitismCount,
                            int tournamentSize) {

        this.populationSize = populationSize;
        this.mutationRate = mutationRate;
        this.crossoverRate = crossoverRate;
        this.elitismCount = elitismCount;
        this.tournamentSize = tournamentSize;
    }

    public Population initPopulation(int chromosomeLength){
        // Initialize population
        Population population = new Population(this.populationSize, chromosomeLength);
        return population;
    }


    public boolean isTerminationConditionMet(int generationsCount, int maxGenerations) {
        return (generationsCount > maxGenerations);
    }

    public Individual selectParent(Population population) {
        // Create tournament
        Population tournament = new Population(this.tournamentSize);

        // Add random individuals to the tournament
        population.shuffle();
        for (int i = 0; i < this.tournamentSize; i++) {
            Individual tournamentIndividual = population.getIndividual(i);
            tournament.setIndividual(i, tournamentIndividual);
        }

        // Return the best
        return tournament.getFittest(0);
    }


    public double calcFitness(Individual individual, City cities[]) {
        Route route = new Route(individual, cities);

        double fitness = 1 / route.getDistance();

        individual.setFitness(fitness);

        return fitness;
    }

    public void evalPopulation(Population population, City cities[]){
        double populationFitness = 0;

        // Loop over population evaluating individuals and summing population fitness
        for (Individual individual : population.getIndividuals()) {
            populationFitness += this.calcFitness(individual, cities);
        }

        double avgFitness = populationFitness / population.size();
        population.setPopulationFitness(avgFitness);
    }

    public Population crossoverPopulation(Population population) {
        // Create new population
        Population newPopulation = new Population(population.size());

        // Loop over current population by fitness
        for (int populationIndex = 0; populationIndex < population.size(); populationIndex++) {
            // Get parent1
            Individual parent1 = population.getFittest(populationIndex);

            // Apply crossover to this individual?
            if (this.crossoverRate > Math.random() && populationIndex >= this.elitismCount) {
                // Find parent2 with tournament selection
                Individual parent2 = this.selectParent(population);

                // Create blank offspring chromosome
                int offspringChromosome[] = new int[parent1.getChromosomeLength()];

                Arrays.fill(offspringChromosome, -1);
                Individual offspring = new Individual(offspringChromosome);

                // Get subset of parent chromosomes

                int substrPos1 = (int) (Math.random() * parent1.getChromosomeLength());

                int substrPos2 = (int) (Math.random() * parent1.getChromosomeLength());

                // make the smaller the start and the larger the end
                final int starSubstr = Math.min(substrPos1, substrPos2);
                final int endSubstr = Math.max(substrPos1, substrPos2);

                // Loop and add the sub tour from parent1 to child

                for (int i = starSubstr; i < endSubstr; i++) {
                    offspring.setGene(i, parent1.getGene(i));
                }

                // Loop through parent2's city tour

                for (int i = 0; i < parent2.getChromosomeLength(); i++) {
                    int parent2Gene = (i + endSubstr) % (parent2.getChromosomeLength());

                    // If offspring doesn't have the city add it
                    if (!offspring.containGene(parent2.getGene(parent2Gene))) {
                        // Loop to find a spare position in the child's tour
                        for (int ii = 0; ii < offspring.getChromosomeLength(); ii++) {
                            // Spare position found, add city
                            if (offspring.getGene(ii) == -1) {
                                offspring.setGene(ii, parent2.getGene(parent2Gene));
                                break;
                            }
                        }
                    }
                }
                newPopulation.setIndividual(populationIndex, offspring);
            } else {
                // Add individual to new population without applying crossover
                newPopulation.setIndividual(populationIndex, parent1);
            }
        }
        return newPopulation;
    }

    public Population mutatePopulation(Population population) {
        // Initialize new population
        Population newPopulation = new Population(this.populationSize);

        // Loop over current population by fitness
        for (int populationIndex = 0; populationIndex < population.size(); populationIndex++) {
            Individual individual = population.getFittest(populationIndex);

            // Skip mutation if this is an elite individual
            if (populationIndex >= this.elitismCount) {
                // Loop over individual's gene
                for (int geneIndex = 0; geneIndex < individual.getChromosomeLength(); geneIndex++) {
                    // Dose this gene need mutation?
                    if (this.mutationRate > Math.random()) {
                        // Get new gene position
                        int newGenePos = (int) (Math.random() * individual.getChromosomeLength());
                        // Get genes to swap
                        int gene1 = individual.getGene(newGenePos);
                        int gene2 = individual.getGene(geneIndex);

                        // Swap genes
                        individual.setGene(geneIndex, gene1);
                        individual.setGene(newGenePos, gene2);
                    }
                }
            }

            // Add individual to population
            newPopulation.setIndividual(populationIndex, individual);
        }
        return newPopulation;
    }
}

File: code/exp1/GeneticAlgorithm.m
classdef GeneticAlgorithm
    properties
        population;
        crossoverRate;
        mutationRate;
        populationSize;
        chromosomeLength;
        fitness;
        tournamentSize;
    end
    
    methods
        function obj = GeneticAlgorithm(population, crossoverRate, mutationRate, fitness, tournamentSize)
            obj.population = population;
            obj.crossoverRate = crossoverRate;
            obj.mutationRate = mutationRate;
            obj.populationSize = size(population, 1);
            obj.chromosomeLength = size(population, 2);
            obj.fitness = fitness;
            obj.tournamentSize = tournamentSize;
        end
        
        function parentIndex = select(obj)
            % candidate store the index of selected chromosome
            candidate = randperm(obj.populationSize, obj.tournamentSize);
            bestIndex = 1;
            for i = 2 : obj.tournamentSize
                if obj.fitness(candidate(i)) < obj.fitness(candidate(bestIndex))
                    bestIndex = i;
                end
            end
            parentIndex = candidate(bestIndex);
        end
        
        function [individual, bestFitness] = selectBestIndividual(obj)
            bestIndex = 1;
            for index = 2 : obj.populationSize
                if obj.fitness(index) < obj.fitness(bestIndex)
                    bestIndex = index;
                end
            end
            individual = obj.population(bestIndex, :);
            bestFitness = obj.fitness(bestIndex);
        end
        
        function newOffspring = mutation(obj, offspring)
            newOffspring = offspring;
            for i = 1 : obj.chromosomeLength
                if rand() < obj.mutationRate
                    newGenePos = randperm(obj.chromosomeLength, 1);
                    gene = newOffspring(newGenePos);
                    
                    newOffspring(newGenePos) = newOffspring(i);
                    newOffspring(i) = gene;
                end
            end
        end
        
        function offspring = crossover(obj,fatherIndex, motherIndex)
            father = obj.population(fatherIndex, :);
            mother = obj.population(motherIndex, :);
            if rand() < obj.crossoverRate
                offspring = zeros(1, obj.chromosomeLength);
                % get subset of parent chromosomes
                substrPos1 = randperm(obj.chromosomeLength, 1);
                substrPos2 = randperm(obj.chromosomeLength, 1);
                
                % make the smaller the start and the larger the end
                startSubstr = min([substrPos1 substrPos2]);
                endSubstr = max([substrPos1 substrPos2]);
                
                offspring(startSubstr : endSubstr) = father(startSubstr : endSubstr);
                
                empty = [1 : startSubstr - 1, endSubstr + 1 : obj.chromosomeLength];
                emptyIndex = 1;
                % Loop through parent2's city tour and add city
                for index = 1 : obj.chromosomeLength
                    geneIndex = index + endSubstr;
                    if geneIndex > obj.chromosomeLength
                        geneIndex = geneIndex - obj.chromosomeLength;
                    end    
                    
                    
                    % if offspring doesn't have the city add it
                    if ~ismember(mother(geneIndex), offspring)
                        offspring(empty(emptyIndex)) = mother(geneIndex);
                        emptyIndex = emptyIndex + 1;
                    end
                end
            else 
                if obj.fitness(fatherIndex) < obj.fitness(motherIndex)
                    offspring = father;
                else 
                    offspring = mother;
                end
            end
        end
    end
end